{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "NOTE: Before running, __go to the last cell__ and run that first."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The Excel file loaded in the next line is from L1Base (v2), a database containing active LINE-1 (L1) insertions residing in human genomes. (http://l1base.charite.de ). The 'hot' L1s here are referenced to hg38. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn\n",
    "plot_dict = {} # for the histogram plot at the end of this notebook. initialized once."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "l1base_df = pd.read_excel('../../data/l1base/l1base_hsfli_8438.xlsx')\n",
    "\n",
    "l1base_list = []\n",
    "for index, row in l1base_df.iterrows():\n",
    "    l1base_list.append([row[0], row[1], row[2], row[5], row[6], row[7], row[8], row[9]])\n",
    "\n",
    "headers = ['Chr','Start','End','Strand','ORF1 Start','ORF1 End','ORF2 Start','ORF2 End']\n",
    "l1base_df = pd.DataFrame(l1base_list, columns=headers)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The function below slices the contents of the Excel file, by chromosome. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def slice_df_by_chrom(df, fields):\n",
    "    '''\n",
    "    Slice a dataframe by chromosome.\n",
    "\n",
    "    :param probe: A dataframe object loaded from l1base_hsfli_8438.xlsx file.\n",
    "    :return: A dictionary of L1 locations keyed by chromosome. \n",
    "    '''\n",
    "    chr_positions = {}\n",
    "    for chr in [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,20,22,'X','Y']: # chr number\n",
    "        chr_positions[chr] = df.loc[df['Chr']==chr][fields] # slice df\n",
    "    return chr_positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Active L1s in chromosome X: \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>ORF1 Start</th>\n",
       "      <th>ORF1 End</th>\n",
       "      <th>ORF2 Start</th>\n",
       "      <th>ORF2 End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>11706247</td>\n",
       "      <td>11714295</td>\n",
       "      <td>1908</td>\n",
       "      <td>2921</td>\n",
       "      <td>2988</td>\n",
       "      <td>6812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>23237515</td>\n",
       "      <td>23245560</td>\n",
       "      <td>1907</td>\n",
       "      <td>2920</td>\n",
       "      <td>2987</td>\n",
       "      <td>6811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>142476843</td>\n",
       "      <td>142484883</td>\n",
       "      <td>1909</td>\n",
       "      <td>2919</td>\n",
       "      <td>2986</td>\n",
       "      <td>6810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>155523050</td>\n",
       "      <td>155515003</td>\n",
       "      <td>1909</td>\n",
       "      <td>2922</td>\n",
       "      <td>2989</td>\n",
       "      <td>6813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>141428248</td>\n",
       "      <td>141420202</td>\n",
       "      <td>1908</td>\n",
       "      <td>2921</td>\n",
       "      <td>2988</td>\n",
       "      <td>6812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>96064845</td>\n",
       "      <td>96056798</td>\n",
       "      <td>1909</td>\n",
       "      <td>2922</td>\n",
       "      <td>2989</td>\n",
       "      <td>6813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>83066639</td>\n",
       "      <td>83058591</td>\n",
       "      <td>1909</td>\n",
       "      <td>2922</td>\n",
       "      <td>2989</td>\n",
       "      <td>6813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>64020288</td>\n",
       "      <td>64012236</td>\n",
       "      <td>1909</td>\n",
       "      <td>2922</td>\n",
       "      <td>2989</td>\n",
       "      <td>6813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>54125747</td>\n",
       "      <td>54117700</td>\n",
       "      <td>1909</td>\n",
       "      <td>2922</td>\n",
       "      <td>2989</td>\n",
       "      <td>6813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>47790701</td>\n",
       "      <td>47782657</td>\n",
       "      <td>1909</td>\n",
       "      <td>2922</td>\n",
       "      <td>2989</td>\n",
       "      <td>6810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>36472219</td>\n",
       "      <td>36464177</td>\n",
       "      <td>1910</td>\n",
       "      <td>2923</td>\n",
       "      <td>2990</td>\n",
       "      <td>6805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>34254915</td>\n",
       "      <td>34247263</td>\n",
       "      <td>1910</td>\n",
       "      <td>2923</td>\n",
       "      <td>2990</td>\n",
       "      <td>6418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>11942317</td>\n",
       "      <td>11934270</td>\n",
       "      <td>1909</td>\n",
       "      <td>2922</td>\n",
       "      <td>2989</td>\n",
       "      <td>6813</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Start        End  ORF1 Start  ORF1 End  ORF2 Start  ORF2 End\n",
       "132   11706247   11714295        1908      2921        2988      6812\n",
       "133   23237515   23245560        1907      2920        2987      6811\n",
       "134  142476843  142484883        1909      2919        2986      6810\n",
       "135  155523050  155515003        1909      2922        2989      6813\n",
       "136  141428248  141420202        1908      2921        2988      6812\n",
       "137   96064845   96056798        1909      2922        2989      6813\n",
       "138   83066639   83058591        1909      2922        2989      6813\n",
       "139   64020288   64012236        1909      2922        2989      6813\n",
       "140   54125747   54117700        1909      2922        2989      6813\n",
       "141   47790701   47782657        1909      2922        2989      6810\n",
       "142   36472219   36464177        1910      2923        2990      6805\n",
       "143   34254915   34247263        1910      2923        2990      6418\n",
       "144   11942317   11934270        1909      2922        2989      6813"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# include the following headings in the sliced object\n",
    "headings = ['Start', 'End', 'ORF1 Start', 'ORF1 End', 'ORF2 Start', 'ORF2 End']\n",
    "\n",
    "chr_dict = slice_df_by_chrom(l1base_df, headings)\n",
    "\n",
    "ch='X' # which chromosome? ----- Note: this sets 'ch' for the entire noteboook. \n",
    "\n",
    "print ('Active L1s in chromosome {n}: '.format(n=ch))\n",
    "L1chr = chr_dict[ch]\n",
    "L1chr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load the methyl object for TCGA-33-4586. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "loading TCGA-50-5931-LUAD ...\n"
     ]
    }
   ],
   "source": [
    "obj = 'TCGA-50-5931-LUAD'\n",
    "TCGA_object = load_obj(obj)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6250"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "probe_keys = sorted(TCGA_object.keys()) # probe keys are 'cg00000029', etc.\n",
    "\n",
    "chr_probes = [] # stores a list of probe keys by chromosome. \n",
    "\n",
    "for probe in probe_keys:\n",
    "    if TCGA_object[probe].chr == str(ch):\n",
    "        chr_probes.append(probe)\n",
    "len(chr_probes) # number of probes belonging to this chromosome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def minmax(val_list):\n",
    "    min_val = min(val_list)\n",
    "    max_val = max(val_list)\n",
    "\n",
    "    return (min_val, max_val)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "L1 = sorted([TCGA_object[probe].hg38_cord for probe in chr_probes]) # loci of TCGA methyl object\n",
    "L2 = sorted(L1chr['End']) # end loci of each LINE-1 (or mid point between start and end?) \n",
    "\n",
    "L1 = list(filter((-1).__ne__, L1)) # remove -1's, WTF?\n",
    "\n",
    "L1_minmax = minmax(L1) # a tuple containing the min and max of the TCGA methyl object.\n",
    "L2_minmax = minmax(L2) # a tuple containing the min and max of the LINE-1 object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "L1: (2796977, 155616255), L2: (11714295, 155515003)\n"
     ]
    }
   ],
   "source": [
    "print ( 'L1: {_L1}, L2: {_L2}'.format(_L1=L1_minmax, \n",
    "                                      _L2=L2_minmax) ) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get the index of the next number between minmax of L1 that is greater than or equal to min of L2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index: 217, Locus: 11755348.\n"
     ]
    }
   ],
   "source": [
    "# <expression1> if <condition> else <expression2>\n",
    "\n",
    "for number in L1:\n",
    "    if number < L2[0]:\n",
    "        continue\n",
    "    if number == L2[0]:\n",
    "        print (L1.index(number))\n",
    "        break\n",
    "    else:\n",
    "        if number > L2[0]:\n",
    "            print ('Index: {i}, Locus: {n}.'.format(i=L1.index(number), \n",
    "                                             n=L1[L1.index(number)]))\n",
    "            break"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Starting from the L1 index above, find all probes that are __within range of L2__. First, remove all elements in L1 that are to the left of index 8599 (that are less than the starting index in L2).  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(217, 6201)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "L1.index(number), L1.index(L1[-1]) # elements in l1 that are in range. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "indexes = range(L1.index(number), L1.index(L1[-1])) # a range of sorted probes in L1\n",
    "\n",
    "new_indexes = []\n",
    "for idx in indexes:\n",
    "    if L1[idx] >= minmax(L2)[1]: # probe index value is greater than greatest of L1's?\n",
    "        break\n",
    "    new_indexes.append(L1[idx])  # if not, add it to the new list. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5976 methylated probes (95.62%) overlap active LINE-1 loci in ChrX for TCGA-50-5931-LUAD.\n"
     ]
    }
   ],
   "source": [
    "\n",
    "try:\n",
    "    \n",
    "    assert (new_indexes[0] > minmax(L2)[0] and new_indexes[-1] < minmax(L2)[1])\n",
    "    \n",
    "    ratio = ((len(new_indexes) / len(chr_probes)) * 100)\n",
    "    ratio = round(ratio, 2)\n",
    "    plot_dict[ch] = ratio\n",
    "    print ('{nl} methylated probes ({r}%) overlap active LINE-1 loci in Chr{c} for {o}.'.format(\n",
    "                                            nl=len(new_indexes),\n",
    "                                            r=ratio,\n",
    "                                            c=ch,\n",
    "                                            o=obj) )\n",
    "except IndexError:\n",
    "    print ('0 probes (0%) overlap active LINE-1s in Chr{c} for {o}.'.format(\n",
    "                                            c=ch,\n",
    "                                            o=obj) )\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot = {'1': 51.6, '2': 65.6, '3': 79.5, '4': 60.9, '5': 73.1, '6': 87.0,\n",
    " '7': 56.8, '8': 29.4, '9': 30.1, '10': 65.0, '11': 55.7, '12': 49.0,\n",
    " '13': 13.0, '14': 34.7, '15': 54.0, '16': 56.7, '17': 0.0, '18': 47.4,\n",
    " '20': 2.5, '22': 59.4, 'X': 95.6, 'Y': 0.0}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The values plotted below are for one cancer type (LUAD) and represent one paired experiment. However the pattern is similar for other types and experiments."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYgAAAEKCAYAAAAIO8L1AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvOIA7rQAAHH9JREFUeJzt3XmcXFWd9/HPVwKyBUgIYAs4QSaD8jDaxAyCoCBxA5kAKogLE9ky+CjDoiM4Po+Io44IisvjiCEgoMgisskgy7C4jIokoYFggCBrICQ4Iokosv3mj3P6SRlvV5+q6uqqrv6+X696Vd1b99T5dd+q+tW959xzFBGYmZmt6UWdDsDMzLqTE4SZmVVygjAzs0pOEGZmVskJwszMKjlBmJlZJScIMzOr1LYEIeksSSskLapZN1nSdZKW5PtJeb0kfVXSvZJulzS9XXGZmVmZdh5BnA28bY11JwDXR8Q04Pq8DLAXMC3f5gDfaGNcZmZWQO28klrSVODKiNghL98N7BERyyT1ATdFxHaSvpkfn7/mdvVef8qUKTF16tS2xW9m1osWLFjwm4jYbLjtJoxGMDW2GPzSz0li87x+S+Dhmu2W5nV/kSAkzSEdZdDX18e8efPaG7GZWY/ZcccdHyzZbrQTxFBUsa7y0CYi5gJzAWbMmBH9/f3tjMvMbNwa7V5My/OpJfL9irx+KbB1zXZbAY+OcmxmZlZjtBPEFcDs/Hg2cHnN+n/IvZl2Bp4crv3BzMzaq22nmCSdD+wBTJG0FDgR+DxwkaTDgIeAA/LmVwF7A/cCfwAOaVdcZmZWpm0JIiLeM8RTMyu2DeBD7YrFzMwa5yupzcyskhOEmZlVcoIwM7NKThBmZlapWy6UMzPrGfPOuK9428OPeHkbI2mNjyDMzKySE4SZmVVygjAzs0pOEGZmVskJwszMKjlBmJlZJScIMzOr5ARhZmaVnCDMzKySE4SZmVVygjAzs0pOEGZmVskJwszMKjlBmJlZJScIMzOr5ARhZmaVnCDMzKySZ5SzP9MrM2GZWet8BGFmZpWcIMzMrJIThJmZVXKCMDOzSk4QZmZWyQnCzMwqOUGYmVklJwgzM6vkBGFmZpWcIMzMrNKQQ21IuigiDpR0BxC1TwEREa9qe3RmZtYx9cZiOjrf7zPSlUo6FjiclHjuAA4B+oALgMnAQuDgiHhmpOs2M7MyQ55iiohlNdssj4gHI+JBYAXpKKIpkrYE/gmYERE7AGsBBwEnA6dFxDTgCeCwZuswM7PWlbRBfA94oWb5+byuFROA9SRNANYHlgF7Ahfn588B9muxDjMza0FJgphQe6onP16n2Qoj4hHgVOAhUmJ4ElgA/C4insubLQW2bLYOMzNrXcl8EI9LmhURVwBI2hf4TbMVSpoE7AtsA/yOdDSyV8WmUbEOSXOAOQB9fX0MDAw0G4pV2HDiquJtBwZWtjESs7GrVz5HJQniSOA8SV/Pyw8DB7dQ55uA+yPicQBJlwCvAzaRNCEfRWwFPFpVOCLmAnMBZsyYEf39/S2EMnrGykQ8828pj7O/3xMGmVXplc/RsAkiIn4N7CxpQ0ARUZ4aqz2UX2994I/ATGA+cCPwLlJPptnA5S3WY2ZmLRi2DULSxpK+BNwE3Cjpi5I2brbCiLiZ1Bi9kNTF9UWkI4LjgeMk3QtsCpzZbB1mZta6klNMZwGLgAPz8sHAt4B3NFtpRJwInLjG6vuAnZp9TTMzG1klCWLbiHhnzfJJktwybGbW40q6uf5R0m6DC5J2JbUdmJlZDyvtxXRuTbvDE6RGZDMz62ElvZhuA14taaO8vFLSO4Hb2x2cmZl1TvFw3xGxMiIGr+g4rU3xmJlZl2h2PoimB+szM7OxodkEUTkMhpmZ9Y56EwatOVHQ/38K2KJtEZmZWVeo10g94hMFmZnZ2DFkgsiTA9kYNVYGBzSz7tVsG4SZmfU4JwgzM6vkBGFmZpWaShCSfjjSgZiZWXep1811+lBPAWNjGjczM2tavW6utwA/ovqq6U3aE46ZmXWLegliMfCPEbFkzSckPdy+kMzMrBvUa4P4VJ3njxr5UMzMrJvUu1Du4jrPXdaecMzMrFs01ItJ0pXtCsTMzLpLo91ct2xLFGZm1nUaTRC3tiUKMzPrOkMmiMEpRmtFxKH5uZe1MygzM+u8ekcQNw0+kHT9Gs+5kdrMrMfVSxC1F8hNrvOcmZn1oHoJIoZ4XLVsZmY9pt6V1JtLOo50tDD4mLy8Wdsj61KeiMfMxot6CeIMYGLFY4B5bYvIrI2c4M3K1buS+qTRDMTMzLpLveG+P1mnXETEv7YhHjMz6xL1TjE9VbFuA+AwYFPACcLMrIfVO8X0xcHHkiYCRwOHABcAXxyqnJmZ9YZ6RxBImgwcB7wPOAeYHhFPjEZgZmbWWfXaIE4B3gHMBf42In4/alGZmVnH1btQ7iPAS4H/AzwqaWW+rZK0spVKJW0i6WJJd0laLGkXSZMlXSdpSb6f1EodZmbWmiETRES8KCLWi4iJEbFRzW1iRPzFQH4N+gpwdUS8Ang1aXrTE4DrI2IacH1eNjOzDql3imnN8Zf+TET8tpkK8yixbwA+kF/nGeAZSfsCe+TNziENFnh8M3X0El/YZWadUq+RegFpzKWqgfkCaPbb6OXA48C3JL0613M0sEVELAOIiGWSNm/y9c3MbATU6+a6zVDPSWplZrkJwHTgqIi4WdJXaOB0kqQ5wByAvr4+BgYGWgilcRtOXFW87cDA6qaaXi83VvT632fdoVfeZ3W7udbxc6DZSYOWAksj4ua8fDEpQSyX1JePHvqAFVWFI2IuqWcVM2bMiP7+/ibDaM78W8pP+fT3rz7I6vVyY0Wv/302spo9xdsr77NGpxwd1PR8EBHxGPCwpO3yqpnAr4ArgNl53Wzg8mbrMDOz1jV7BNHqfBBHAedJWge4j3SF9ouAiyQdBjwEHNBiHWZm1oJ6vZi+RnUiELBJK5VGxAAwo+Kpma28rpmZjZx6RxDzm3zOzMx6QL1eTOeMZiBmZtZdmm2DMPszvqBvfPJ+723N9mIyM7Me5yMIswL+pWzj0bAJQtJmwBHA1NrtI+LQ9oVlZmadVnIEcTnwE+A/gefbG46ZmXWLkgSxfkSM+1FVzczGm5IEcaWkvSPiqrZHY2Yd4TYWq1LvSupVrB7u+18k/Ql4Ni/HCEwaZGZmXazehXITRzMQMzPrLsNeByHp+pJ1ZmbWW+qdYloX2ACYImkSq4f43gh46SjEZmZmHVSvkfofgWNIyWBhzfqVwNfbGdRocKOcmVl99dogvgJ8RdJREfG1UYzJzMy6QL1TTHtGxA3AI5LesebzEXFJWyMzM7OOqneKaXfgBuDvK54LwAnCzKyH1TvFdGK+P2T0wjEzs25RMljfr4FfkMZj+nFE/KrtUZmZWceVDLWxPfBa4PXAqZJeAdwWEfu3NTKzHuDecjaWlUwY9DxpiI3ngReA5cCKdgZlZmadV3IEsRK4A/gScEZE/Hd7QzIzs25QkiDeA+wG/G/gcEk/I7VFeLgNMxtVPmU3uoZNEBFxOXB5bnvYi3R19ceA9docm5mZdVDJYH3fzz2ZvkIam+kfgEntDszMzDqr5BTT54GFEeHpRs3MxpGSU0y3jEYgZmbWXUq6uZqZ2TjkBGFmZpVKGqkl6f2SPpmXXyZpp/aHZmZmnVRyBPHvwC6k6yEAVtEDEwaZmVl9Jb2YXhsR0yXdChART0hap81xmZlZh5UcQTwraS3SHBBI2ow0JpOZmfWwkgTxVeBSYHNJnwV+CnyurVGZmVnHlVwHcZ6kBcBMQMB+EbG41YrzUcl84JGI2EfSNsAFwGRgIXBwRDzTaj1mZtackl5Mk0nDe58PfBdYLmntEaj7aKA20ZwMnBYR04AngMNGoA4zM2tSySmmhcDjwD3Akvz4fkkLJb2mmUolbQW8HZiXlwXsCVycNzkH2K+Z1zYzs5FRkiCuBvaOiCkRsSlpRNeLSMN//3uT9X6ZNCLsYGP3psDvIuK5vLwU2LLJ1zYzsxFQ0s11RkQcObgQEddK+lxEHCfpxY1WKGkfYEVELJC0x+Dqik1jiPJzgDkAfX19DAwMNBoCABtOXFW87cDASpdrU7nR5v9LtbHy942V+sbKfh9OSYL4raTjSQ3IAO8GnsiNzM10d90VmCVpb2BdYCPSEcUmkibko4itgEerCkfEXGAuwIwZM6K/v7+JEGD+LeUTj/T3r554xOVGttxo8/+l2lj5+8ZKfWNlvw+n5BTTe0lf2JcBlwMvy+vWAg5stMKI+HhEbBURU4GDgBsi4n3AjcC78mazc11mZtYhJd1cfwMcNcTT945gLMcDF0j6DHArcOYIvraZmTVo2ASRr5z+GPC/SKeEAIiIPVutPCJuAm7Kj+8DPAigmVmXKDnFdB5wF7ANcBLwAOBJhMzMelxJgtg0Is4Eno2IH0XEocDObY7LzMw6rKQX07P5fpmkt5N6F23VvpDMzKwblCSIz0jaGPgI8DVSt9Rj2xqVmZl1XN0Eka91mBYRVwJPAm8clajMzKzj6rZBRMTzwKxRisXMzLpIySmmn0n6f8CFwFODKyNiYduiMjOzjitJEK/L95+uWRek0VfNrIvMO6N8iIfDj+jeIR6sO5RcSe12BzOzcahkwqAtJJ0p6Yd5eXtJnszHzKzHlVwodzZwDfDSvHwPcEy7AjIzs+5QkiCmRMRF5KG983Dcz7c1KjMz67iSBPGUpE3JE/hI2pl0TYSZmfWwkl5MHwGuALaV9F/AZqyet8HMzHpUSS+mBZJ2B7YjTQ16d0Q8O0wxMzMb40p6Md1Gmg/i6YhY5ORgZjY+lLRBzAKeAy6SdIukj0p6WZvjMjOzDis5xfQg8AXgC5KmAf8XOJk0J7VZR/iKYbP2K2mkRtJU4EDg3aQurh9rX0g2nviL3qx7lcxJfTOwNvA94IA8d7SZmfW4kiOI2RFxV9sjMTOzrlLSSL1M0pckzc+3L+YZ5szMrIeVJIizgFWkNogDgZXAt9oZlJmZdV7JKaZtI+KdNcsnSRpoV0BmZtYdSo4g/ihpt8EFSbsCf2xfSGZm1g1KjiCOBM6taXd4AvhA2yIyM7OuUHKh3G3AqyVtlJdXtj0qMzPruCFPMUk6rnbmuIhYGRErJR0lyRMGmZn1uHptEIcC365YPzc/Z2ZmPaxegoiIeKZi5Z9Iw36bmVkPq9uLSdIWJevMzKz31EsQpwD/IWl3SRPzbQ/gB8CpoxKdmZl1zJC9mCLiXEmPA58GdiDNSX0ncGJE/HCU4jMzsw6p2801JwInAzOzcajkSmozMxuHRj1BSNpa0o2SFku6U9LRef1kSddJWpLvJ412bGZmtlonjiCeAz4SEa8EdgY+JGl74ATg+oiYBlyfl83MrEOKE4SknSXdIOm/JO3XbIURsSwiFubHq4DFwJbAvsA5ebNzgKbrMDOz1g3ZSC3pJRHxWM2q44BZpIvkfgZc1mrlea7rHYGbgS0iYhmkJCJp8yHKzAHmAPT19TEw0NzI4xtOXFW87cDA6uGnXM7lRqNcs/z3dUd9ox1nu9TrxXS6pAXAKRHxNPA74L3AC6RJg1oiaUPg+8AxeYynonIRMZc03AczZsyI/v7+puqff0v51Nr9/S93OZcb1XLN8t/XHfWNdpztMuQppojYDxgArpR0MHAMKTmsT4unfyStTUoO50XEJXn1ckl9+fk+YEUrdZiZWWvqtkFExA+AtwKbAJcAd0fEVyPi8WYrVDpUOBNYHBFfqnnqCmB2fjwbuLzZOszMrHX1hvueJemnwA3AIuAgYH9J50vatoU6dwUOBvaUNJBvewOfB94saQnw5rxsZmYdUq8N4jPALsB6wFURsRNwnKRpwGdJCaNhEfFThh4NdmYzr2lmZiOvXoJ4kpQE1qOmPSAiltBkcjAzs7GjXhvE/qQG6edIvZfMzGwcqTea62+Ar41iLGZm1kU8WJ+ZmVVygjAzs0pOEGZmVskJwszMKjlBmJlZJScIMzOr5ARhZmaV6l1JbWYdMu+M8uGiDz+ie4eLtrHNRxBmZlbJCcLMzCo5QZiZWSUnCDMzq+QEYWZmlZwgzMyskhOEmZlVcoIwM7NKThBmZlbJCcLMzCo5QZiZWSUnCDMzq+QEYWZmlZwgzMyskhOEmZlVcoIwM7NKThBmZlbJCcLMzCo5QZiZWSUnCDMzq+QEYWZmlSZ0OgAzs3abd8Z9xdsefsTL2xjJ2OIjCDMzq9RVCULS2yTdLeleSSd0Oh4zs/GsaxKEpLWArwN7AdsD75G0fWejMjMbv7omQQA7AfdGxH0R8QxwAbBvh2MyMxu3uilBbAk8XLO8NK8zM7MOUER0OgYAJB0AvDUiDs/LBwM7RcRRa2w3B5iTF7cD7h7hUKYAv3E5l+uyulzO5UbSX0XEZsNuFRFdcQN2Aa6pWf448PEOxDHf5Vyu2+pyOZfrxK2bTjHdAkyTtI2kdYCDgCs6HJOZ2bjVNRfKRcRzkj4MXAOsBZwVEXd2OCwzs3GraxIEQERcBVzV4TDmupzLdWFdLudyo65rGqnNzKy7dFMbhJmZdREniEzSWZJWSFrUQJmtJd0oabGkOyUdXVhuXUm/lHRbLndSg7GuJelWSVc2UOYBSXdIGpA0v4Fym0i6WNJd+e/cpaDMdrmewdtKSccU1nds/p8sknS+pHULyx2dy9xZr66q/SxpsqTrJC3J95MKyx2Q63tB0owG6jsl/z9vl3SppE0Ky/1rLjMg6VpJLy0pV/PcRyWFpCmF9X1K0iM1+3Hv0vokHZWHzblT0hcK67uwpq4HJA0UluuX9IvB97aknSrKVX5Wh9v3dcoNuw9ryt8vaXJenpSX/6pq+5pykvRTSXvVrDtQ0tX1yo24TnWf6rYb8AZgOrCogTJ9wPT8eCJwD7B9QTkBG+bHawM3Azs3UO9xwHeBKxso8wAwpYn/yznA4fnxOsAmDZZfC3iM1O96uG23BO4H1svLFwEfKCi3A7AIWJ/UrvafwLTS/Qx8ATghPz4BOLmw3CtJ1+LcBMxooL63ABPy45MbqG+jmsf/BJxe+j4GtiZ1AHmw6n0wRH2fAj46zP++qtwb8z54cV7evDTOmue/CHyysL5rgb3y472BmyrKVX5Wh9v3dcoNuw9rXuNjwNz8+JsUdt/P7+vFwLrABsASYNtGPn+t3nwEkUXEj4HfNlhmWUQszI9XkXbmsFd/R/L7vLh2vhU1BknaCng7MK+RWJshaSPSB/JMgIh4JiJ+1+DLzAR+HREPFm4/AVhP0gTSF/6jBWVeCfwiIv4QEc8BPwL2r9pwiP28LykRku/3KykXEYsjou6FmkOUuzbHCfALYKvCcitrFjeg4j1T5318GumLqvJ91sz7v065DwKfj4g/5W1WNFKfJAEHAucXlgtgo/x4YyreM3U+q3X3/VDlSvZhjdOAnfOR7W6k5DesiFgE/AA4HjgRODcifl1SdqQ4QYwQSVOBHUlHAyXbr5UPoVcA10VEUTngy6QP+gsNhhjAtZIWKF2NXuLlwOPAt5ROac2TtEGD9R5ExQe9MsCIR4BTgYeAZcCTEXFtQdFFwBskbSppfdKvyK0biHGLiFiWY1gGbN5A2VYdCvywdGNJn5X0MPA+4JOFZWYBj0TEbU3E9+F8GuWsqlNvQ/gb4PWSbpb0I0l/12CdrweWR8SSwu2PAU7J/5dTSRfZDmmNz2rxvq/zGa+7DyPiWeCfSYnimEhjzZU6CXgvaRDTvzhV125OECNA0obA90k7f+Vw2wNExPMR0U/65bGTpB0K6tkHWBERC5oIc9eImE56o31I0hsKykwgHc5/IyJ2BJ4iHYYXUbrgcRbwvcLtJ5F+0W0DvBTYQNL7hysXEYtJh/nXAVcDtwHP1S3UBSR9ghTneaVlIuITEbF1LvPhgjrWBz5BYTJZwzeAbYF+UsIu+uVLet9MAnYmfTFelI8KSr2Hwh8V2QeBY/P/5VjyEW+VZj6r9co1sA/3Iv0Ph/2c14qIp4ALgW8PHpGNJieIFklam/TGOS8iLmm0fD5lcxPwtoLNdwVmSXqANNrtnpK+U1jPo/l+BXApafTc4SwFltYc3VxMShil9gIWRsTywu3fBNwfEY/nX12XAK8rKRgRZ0bE9Ih4A+kUROmvT4DlkvoA8v1fnBIZaZJmA/sA74t8wrlB3wXeWbDdtqSEe1t+32wFLJT0kuEKRsTy/EPmBeAMyt4zkN43l+RTqb8kHe3+RcN4lXxq8R2kL8VSs0nvFUg/RirjHOKzOuy+H+ozXroPJfUDbyYlzGMH62vACzR+xmBEOEG0IP8qOhNYHBFfaqDcZoO9HiStR/pivGu4chHx8YjYKiKmkk7d3BARw/7ClrSBpImDj0kNbMP21oqIx4CHJW2XV80EfjVcuRqN/hJ8iHSudv38v51JOuc7LEmb5/uXkb5gGqn3CtKXDPn+8gbKNkzS20jnlWdFxB8aKDetZnEWZe+ZOyJi84iYmt83S0mNro8V1Ff7RbY/Be+Z7DJgz/waf0Pq3FA62NybgLsiYmnh9pDaHHbPj/ek4sdBnc9q3X0/VLnSfZjLf4N05PEQcArpNNjYMNKt3mP1RvpCWQY8S/oQHVZQZjfSuf3bgYF827ug3KuAW3O5RVT01ih4jT0o7MVEaku4Ld/uBD7RQD39wPwc62XApMJy6wP/DWzc4N91EumLbxHwbXJPmIJyPyElr9uAmY3sZ2BT4HrSF8v1wOTCcvvnx38CllMz2OQw5e4lDW0/+J6p6o1UVe77+f9yO6nxcstG38cM0ZttiPq+DdyR67sC6Csstw7wnRzrQmDP0jiBs4EjG9x/uwEL8r6/GXhN6Wd1uH1fp9yw+zCXnwNcWLO8Vo519wY+E59imN5k7br5SmozM6vkU0xmZlbJCcLMzCo5QZiZWSUnCDMzq+QEYWZmlZwgrGdJeomkCyT9WtKvJF0laY4aGAXXbDxzgrCelC9QupQ0sue2EbE98C/AFoXl12pnfGZjgROE9ao3As9GxOmDKyJigHRB3YZaPcfFeYPjBCnNQfBJST8FDtDqeQYGx/yflLe7SdJpkn6sNE/A30m6RGlOgc8M1ifpOKU5KhblkTwHr2r/D6W5QBZJendePzMPiHhHHhjvxTUxfU7Sz5XmOpgu6Zp8VHRkTV3/LOmWHGtD84uYDcUJwnrVDqQrVqvsSBoBdHvSVea71jz3dETsFhEXAOcCx0fEq0hXFJ9Ys90zkcZ9Op00PMOHcp0fyKPKvgY4BHgtaQyeIyTtSBpz69GIeHVE7ABcrTQp0tnAuyPib0mD3X2wpq6HI2IXUnI7G3hXfs1PA0h6CzCNNAZRP/CawsEYzepygrDx6JcRsTTSIHQDwNSa5y4EkLQxaXKkH+X155Dmxhh0Rb6/A7gz0rwBfwLuIw01vhtwaUQ8FWnuj0tIw1jfAbxJ0smSXh8RT5ImHbo/Iu4pqOvmiFgVEY8DT+cxvd6Sb7eShrZ4BSlhmLVkQqcDMGuTO0m/tKvUDpv8PH/+OXiq8PUHX+OFNV7vhfx6lcNbR8Q9+ehib+DfJF3L6gTQSl3/FhHfLIzdrIiPIKxX3QC8WNIRgyuUJq7Zfegiq+Vf9k9Ien1edTBpprpSPwb2yyPTbkAa2O8nSvNI/yEivkMa1XM6aXDCqZL+usm6rgEOzXMWIGnLwdFtzVrhIwjrSRERkvYHvizpBOBp0kimlzXwMrOB0/OkO/eR2hRK618o6Wzgl3nVvIi4VdJbSbOfvUAakfSDEfG0pEOA7+X5EG4htW2U1nWtpFcCP8/t7b8H3s8ozGthvc2juZqZWSWfYjIzs0pOEGZmVskJwszMKjlBmJlZJScIMzOr5ARhZmaVnCDMzKySE4SZmVX6H4bm9lMZnSXRAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt \n",
    "\n",
    "plt.bar(range(len(plot)), list(plot.values()), color='#0504aa', alpha=0.4, align='center')\n",
    "plt.grid(axis='y', alpha=0.75)\n",
    "plt.xticks(range(len(plot)), list(plot.keys()))\n",
    "plt.xlabel('Chromosome')\n",
    "plt.ylabel('% Coverage with LINE-1 Loci')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The plot shows relative coverage of 'methylated' Illumina 450k probes with LINE-1 L1Base data (by chromosome). ~50% of 450k probes for chromosome 1 overlap with loci found in the L1Base data for LINE-1, ~95% overlap for Chr X.   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "content = []\n",
    "eul1db_srip_list = []\n",
    "headers = ['Chr','Start','End', 'Strand']\n",
    "\n",
    "with open(\"../../data/eul1db-data/eul1db_srip.bed\")as f:\n",
    "    for line in f:\n",
    "        content.append(line.strip().split())\n",
    "for i, line in enumerate(content):\n",
    "    if i < 3:\n",
    "        continue\n",
    "    eul1db_srip_list.append([line[0][3:], line[1], line[2], line[5]])\n",
    "eul1db_srip_df = pd.DataFrame(eul1db_srip_list, columns=headers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>815511</td>\n",
       "      <td>815511</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>980069</td>\n",
       "      <td>980069</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1149273</td>\n",
       "      <td>1149273</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1687509</td>\n",
       "      <td>1687509</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>3053277</td>\n",
       "      <td>3053277</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chr    Start      End Strand\n",
       "0   1   815511   815511      -\n",
       "1   1   980069   980069      +\n",
       "2   1  1149273  1149273      -\n",
       "3   1  1687509  1687509      -\n",
       "4   1  3053277  3053277      -"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "eul1db_srip_df.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %load '../methylator/annotation/annotate_450k.py'\n",
    "import os\n",
    "\n",
    "class Probe:\n",
    "    \"\"\"\n",
    "    Holds Illumina 450k probe info for a single CpG site.\n",
    "    \"\"\"\n",
    "    def __init__(self):\n",
    "        self.id = None\n",
    "        self.seq = None\n",
    "        self.name = None\n",
    "        self.chr = None\n",
    "        self.hg19_cord = None\n",
    "        self.strand = None\n",
    "        self.gene = None\n",
    "        self.refseq = None\n",
    "        self.tour = None\n",
    "        self.feature = None\n",
    "        \n",
    "class Interval:\n",
    "    \"\"\"\n",
    "    Define a genomic interval by chromsome and strand orientation.\n",
    "    \"\"\"\n",
    "    def __init__(self, chromosome, start, end, strand):\n",
    "        self.chr = chromosome\n",
    "        self.start = start\n",
    "        self.end = end\n",
    "        self.strand = strand\n",
    "\n",
    "class Feature:\n",
    "    \"\"\"\n",
    "    Define a Probe feature.\n",
    "    \"\"\"\n",
    "    BODY = \"Body\"\n",
    "    TSS200 = \"TSS200\"\n",
    "    TSS1500 = \"TSS1500\"\n",
    "    UTR5 = \"5'UTR\"\n",
    "    UTR3 = \"3'UTR\"\n",
    "    EXON = \"Exon\"\n",
    "\n",
    "class CpG_location:\n",
    "    \"\"\"\n",
    "    Defines a CpG location.\n",
    "    \"\"\"\n",
    "    ISLAND = \"Island\"\n",
    "    NSHORE = \"N_Shore\"\n",
    "    SSHORE = \"S_Shore\"\n",
    "    NSHELF = \"N_Shelf\"\n",
    "    SSHELF = \"S_Shelf\"\n",
    "    \n",
    "class SNP:\n",
    "    \"\"\"\n",
    "    Defines the SNPs in probes. Used to filter probes.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self):\n",
    "        self.probeid = None\n",
    "        self.snpid = None\n",
    "    \n",
    "class Annotate_450k:\n",
    "    \"\"\"\n",
    "    Parse and hold information about Illumina probes.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self):        \n",
    "        for i in open(anno_file, mode=\"r\"):\n",
    "            self.ann = os.path.join(\"../../data/\", i.strip(\"\\n\").strip(\"\\r\"))\n",
    "\n",
    "        self.probe = {}\n",
    "        self.__run__()\n",
    "\n",
    "    def __run__(self):\n",
    "        \"\"\"\n",
    "        A static function to setup the Probe classes.\n",
    "        \"\"\"\n",
    "        for i in open(self.ann, mode=\"r\"):\n",
    "            if i.startswith(\"cg\"):\n",
    "                data = i.split(\",\")\n",
    "                # Assign probe information.\n",
    "                new_probe = Probe()\n",
    "                new_probe.id = data[0]                 # ID\n",
    "                new_probe.name = data[1]               # name\n",
    "                new_probe.seq = data[13]               # sequence prior to bi-sulfite conv.\n",
    "                new_probe.chr = str(data[11])          # chromosome\n",
    "                new_probe.hg19_cord = int(data[12])    # chromosomal coordinates of the CpG.\n",
    "                new_probe.strand = data[16]            # strand\n",
    "                new_probe.gene = data[21].split(\";\")   # UCSC_RefGene_Name\n",
    "                new_probe.refseq = data[22]            # UCSC_RefGene_Accession\n",
    "                features = data[23].split(\";\")         # UCSC_RefGene_Group\n",
    "                list_features = []\n",
    "                for i in features:\n",
    "                    if i not in list_features:\n",
    "                        list_features.append(i)\n",
    "\n",
    "                new_probe.feature = list_features\n",
    "\n",
    "                new_probe.tour = data[25]              # Relation_to_UCSC_CpG_Island\n",
    "                newcpg = {new_probe.id: new_probe}\n",
    "                self.probe.update(newcpg)\n",
    "\n",
    "    def get_number(self):\n",
    "        \"\"\"\n",
    "        Return total number of probes.\n",
    "        \"\"\"\n",
    "        number = 0\n",
    "        for probe_id in self.probe.keys():\n",
    "            number += 1\n",
    "\n",
    "        return number\n",
    "\n",
    "    def get_probe(self, probe_id): #WORKS\n",
    "        \"\"\"\n",
    "        Return probe info associated with an reference.\n",
    "        \"\"\"\n",
    "        try:\n",
    "            probe = self.probe[probe_id]\n",
    "        except Exception as ex:\n",
    "            probe = None\n",
    "            print(\"WARNING: No probe with ref-id of %s found.\" % probe_id)\n",
    "        return probe\n",
    "\n",
    "    def remove_snp_probes(self):\n",
    "        \"\"\"\n",
    "        Removes all SNPs associated with probes.\n",
    "        \"\"\"\n",
    "        snp_list = []\n",
    "        snp_file = open(\"../../data/humanmethylation450_dbsnp137.snpupdate.table.v2.sorted.txt\", \"r\")\n",
    "        for line in snp_file:\n",
    "            if line.startswith(\"cg\"):\n",
    "                line = line.strip(\"\\n\").strip(\"\\r\").split(\"\\t\")\n",
    "                new_snp = SNP()\n",
    "                new_snp.probeid = line[0]\n",
    "                new_snp.snpid = line[1]\n",
    "                snp_list.append(new_snp)\n",
    "\n",
    "        for snp in snp_list:\n",
    "            self.probe.pop(snp.probeid)\n",
    "\n",
    "anno_file = os.path.abspath(\"../../data/config.ini\") # Illumina probe manifest. Note: This (large) file is not \n",
    "                                                     # in the repository.\n",
    "# Functions to save/load dictionary objects. \n",
    "\n",
    "import _pickle as pickle\n",
    "\n",
    "def save_obj(obj, name):\n",
    "    print('saving ' + name + ' ...')\n",
    "    with open('../../data/pickle/'+ name + '.pkl', 'wb+') as f:\n",
    "        pickle.dump(obj, f)\n",
    "        \n",
    "def load_obj(name):\n",
    "    print('loading ' + name + ' ...')\n",
    "    with open('../../data/pickle/' + name + '.pkl', 'rb') as f:\n",
    "        return pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
