## Papers

##### Papers included in this directory are listed here more or less in the order of their importance. Note: If the paper is not located here that means the reference provided is an HTML resource.

* [The Minfi User Guide](https://bioconductor.org/packages/release/bioc/vignettes/minfi/inst/doc/minfi.html) - 
A comprehensive guide provided by the authors of minfi. 

* [Minfi: a flexible and comprehensive Bioconductor package for the analysis of Infinium DNA methylation microarrays](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4016708/) - Demonstrates methods for preprocessing, quality assessment and detection of differentially methylated regions. An important paper to read. 

* [ A cross-package Bioconductor workflow for analysing methylation array data.](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4916993/) - Provides a workflow for the analysis of methylation array data. Specifically demonstrates the steps involved in a typical differential methylation analysis pipeline. 

* [shinyMethyl: interactive quality control of Illumina 450k DNA methylation arrays in R](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4176427/) - Facilitates QC of methylation datasets available through The Cancer Genome Atlas (TCGA).

* [450k workflow with Minfi - From BioC Conference 2015](https://www.bioconductor.org/help/course-materials/2015/BioC2015/methylation450k.html) - An older review. Parts 1 and 2 of our tutorial are based on. 

* [Analysis of 450k DNA methylation data with Minfi](https://kasperdanielhansen.github.io/genbioconductor/html/minfi.html) - A tutorial provided by minfi author. Demonstrates the use of **GEO** with minfi. 

* [Functions of DNA methylation: islands, start sites, gene bodies and beyond](https://www.researchgate.net/publication/225068812_Functions_of_DNA_methylation_Islands_start_sites_gene_bodies_and_beyond) - An excellent review of the importance of methylation studies in general (nothing to do with analysing the data or with minfi). 

* [Analysis pipelines and packages for Infinium HumanMethylation450 BeadChip (450k) data](http://www.sciencedirect.com/science/article/pii/S1046202314002692) - A review of the most popular pipelines and packages for the Illumina 450k. Aimed at new 450k users.

* [An evaluation of supervised methods for identifying differentially methylated regions in Illumina methylation arrays](https://academic.oup.com/bib/advance-article/doi/10.1093/bib/bby085/5096828) - A comprehensive evaluation of 4 popular DMR analysis tools under 60 different parameter settings.

* [A survey of the approaches for identifying differential methylation using bisulfite sequencing data](http://www.cs.wayne.edu/tinnguyen/papers/MethylationReview.pdf) - A detailed survey of 22 different approaches for identifying differential methylation. Focuses on underlying statistical models, primary features, key advantages and limitations.

* [Nature Genetics review: Statistical and integrative system-level analysis of DNA methylation data](https://www.nature.com/articles/nrg.2017.86) - An in-depth review of methylation profiling studies. Not directly related but interesting.

------
##### Here are papers specific to LINE-1 retrotransposons.

* [Activation of individual L1 retrotransposon instances is restricted to cell-type dependent permissive loci](https://elifesciences.org/articles/13926) - This paper provides an in-depth look at a very restricted subset of L1HS-Ta loci - some being polymorphic among individuals - which significantly contributes to the bulk of L1 expression. These loci are differentially regulated among distinct cell lines. The data presented support a local model of L1 transcriptional activation in somatic cells, governed by individual-, locus-, and cell-type-specific determinants.
