## Papers

##### Included papers are listed here more or less in the order of their importance. Note: If the paper is not located here it's on the shared drive.

##### Papers specific to LINE-1's and tumorigenesis.

* [Activation of individual L1 retrotransposon instances is restricted to cell-type dependent permissive loci](https://elifesciences.org/articles/13926) - This paper provides an in-depth look at a very restricted subset of L1HS-Ta loci - some being polymorphic among individuals - which significantly contributes to the bulk of L1 expression. These loci are differentially regulated among distinct cell lines. The data presented support a local model of L1 transcriptional activation in somatic cells, governed by individual-, locus-, and cell-type-specific determinants.

* [Cancer Genome Landscapes](https://science.sciencemag.org/content/339/6127/1546.full) - Recent studies have revealed ~140 genes that, when altered by intragenic mutations, promote or “drive” tumorigenesis. A typical tumor contains two to eight of these “driver gene” mutations; the remaining mutations are passengers that confer no selective growth advantage. 

* [Minimal functional driver gene heterogeneity among untreated metastases](https://www.ncbi.nlm.nih.gov/pubmed/30190408) - This paper considers sequencing data for 76 untreated metastases from 20 patients and inferred cancer phylogenies for breast, colorectal, endometrial, gastric, lung, melanoma, pancreatic, and prostate cancers. They find that within individual patients, a large majority of driver gene mutations are common to all metastases.

##### Papers specific to DNA methylation.

* [Functions of DNA methylation: islands, start sites, gene bodies and beyond](https://www.researchgate.net/publication/225068812_Functions_of_DNA_methylation_Islands_start_sites_gene_bodies_and_beyond) - An excellent review of the importance of methylation studies in general (nothing to do with analysing the data or with minfi). 

* [A survey of the approaches for identifying differential methylation using bisulfite sequencing data](http://www.cs.wayne.edu/tinnguyen/papers/MethylationReview.pdf) - A detailed survey of 22 different approaches for identifying differential methylation. Focuses on underlying statistical models, primary features, key advantages and limitations.

* [Nature Genetics review: Statistical and integrative system-level analysis of DNA methylation data](https://www.nature.com/articles/nrg.2017.86) - An in-depth review of methylation profiling studies. Not directly related but interesting.

##### Papers specific to analysis methods.

* [Analysis pipelines and packages for Infinium HumanMethylation450 BeadChip (450k) data](http://www.sciencedirect.com/science/article/pii/S1046202314002692) - A review of the most popular pipelines and packages for the Illumina 450k. Aimed at new 450k users.

[CpG Loci Identication](https://www.illumina.com/Documents/products/technotes/technote_cpg_loci_identification.pdf) - A guide to Illumina’s method for unambiguous CpG loci identi cation and tracking for the GoldenGate® and In nium® Assays for Methylation.

* [An evaluation of supervised methods for identifying differentially methylated regions in Illumina methylation arrays](https://academic.oup.com/bib/advance-article/doi/10.1093/bib/bby085/5096828) - A comprehensive evaluation of 4 popular DMR analysis tools under 60 different parameter settings.

##### Papers specific to software.

* [Minfi: a flexible and comprehensive Bioconductor package for the analysis of Infinium DNA methylation microarrays](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4016708/) - Demonstrates methods for preprocessing, quality assessment and detection of differentially methylated regions. An important paper to read. 

* [Analysis of 450k DNA methylation data with Minfi](https://kasperdanielhansen.github.io/genbioconductor/html/minfi.html) - A tutorial provided by minfi author. Demonstrates the use of **GEO** with minfi.