{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "NOTE: If running for the first time uncomment and run the last line of this notebook first."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'../../data/pickle/TCGA-43-6771-LUSC.pkl'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pickle_base = '../../data/pickle/'\n",
    "tissue      = 'LUSC'\n",
    "directory   = tissue + '-PAIRED/'\n",
    "\n",
    "# Paired experiments\n",
    "TCGA   = {\"LUSC\": ['43-6771', '18-5595', '22-5471', '39-5031', '33-4589', '33-4586'],\n",
    "          \"LUAD\": ['50-5931', '15-5420', '38-4631', '49-4488', '50-5932', '50-5935'],\n",
    "          \"COAD\": ['AA-3697', 'AA-3713', 'AA-3506', 'AZ-6599', 'A6-2686', 'A6-2682'],\n",
    "          \"BRCA\": ['BH-A0BZ', 'BH-A1EN', 'BH-A1F5', 'E2-A1IO'],          \n",
    "          \"GBM\":  ['74-6573', '06-0152']}\n",
    "    \n",
    "minor       =  TCGA['LUSC'][0][:2]\n",
    "major       =  TCGA['LUSC'][0][3:]\n",
    "\n",
    "anno_str       = 'Illumina-450k-Anno.{rev}.{ext}'.format(rev='hg19',ext='pkl')\n",
    "file_str_picl  = 'TCGA-{minor}-{major}-{tissue}.{ext}'.format(minor=minor,\n",
    "                                                                major=major,\n",
    "                                                                tissue=tissue,\n",
    "                                                                ext='pkl')\n",
    "\n",
    "filepath = pickle_base + file_str_picl\n",
    "filepath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def slice_df_by_chrom(df, fields):\n",
    "    '''\n",
    "    Slice a dataframe by chromosome.\n",
    "\n",
    "    :param probe: A dataframe object loaded from l1base_hsfli_8438.xlsx file.\n",
    "    :return: A dictionary of L1 locations keyed by chromosome. \n",
    "    '''\n",
    "    chr_positions = {}\n",
    "    for chr in [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,20,22,'X','Y']: # chr number\n",
    "        chr_positions[chr] = df.loc[df['Chr']==chr][fields] # slice df\n",
    "    return chr_positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "anno = Annotate_Methyl_450k(filepath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "l1base_df = pd.read_excel('../../data/l1base/l1base_hsfli_8438.xlsx')\n",
    "\n",
    "l1base_list = []\n",
    "for index, row in l1base_df.iterrows():\n",
    "    l1base_list.append([row[0], row[1], row[2], row[5], row[6], row[7], row[8], row[9]])\n",
    "\n",
    "headers = ['Chr','Start','End','Strand','ORF1 Start','ORF1 End','ORF2 Start','ORF2 End']\n",
    "l1base_df = pd.DataFrame(l1base_list, columns=headers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Active L1s in chromosome X: \n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>ORF1 Start</th>\n",
       "      <th>ORF1 End</th>\n",
       "      <th>ORF2 Start</th>\n",
       "      <th>ORF2 End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>11706247</td>\n",
       "      <td>11714295</td>\n",
       "      <td>1908</td>\n",
       "      <td>2921</td>\n",
       "      <td>2988</td>\n",
       "      <td>6812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>23237515</td>\n",
       "      <td>23245560</td>\n",
       "      <td>1907</td>\n",
       "      <td>2920</td>\n",
       "      <td>2987</td>\n",
       "      <td>6811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>142476843</td>\n",
       "      <td>142484883</td>\n",
       "      <td>1909</td>\n",
       "      <td>2919</td>\n",
       "      <td>2986</td>\n",
       "      <td>6810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>155523050</td>\n",
       "      <td>155515003</td>\n",
       "      <td>1909</td>\n",
       "      <td>2922</td>\n",
       "      <td>2989</td>\n",
       "      <td>6813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>141428248</td>\n",
       "      <td>141420202</td>\n",
       "      <td>1908</td>\n",
       "      <td>2921</td>\n",
       "      <td>2988</td>\n",
       "      <td>6812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>137</th>\n",
       "      <td>96064845</td>\n",
       "      <td>96056798</td>\n",
       "      <td>1909</td>\n",
       "      <td>2922</td>\n",
       "      <td>2989</td>\n",
       "      <td>6813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138</th>\n",
       "      <td>83066639</td>\n",
       "      <td>83058591</td>\n",
       "      <td>1909</td>\n",
       "      <td>2922</td>\n",
       "      <td>2989</td>\n",
       "      <td>6813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>139</th>\n",
       "      <td>64020288</td>\n",
       "      <td>64012236</td>\n",
       "      <td>1909</td>\n",
       "      <td>2922</td>\n",
       "      <td>2989</td>\n",
       "      <td>6813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>140</th>\n",
       "      <td>54125747</td>\n",
       "      <td>54117700</td>\n",
       "      <td>1909</td>\n",
       "      <td>2922</td>\n",
       "      <td>2989</td>\n",
       "      <td>6813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141</th>\n",
       "      <td>47790701</td>\n",
       "      <td>47782657</td>\n",
       "      <td>1909</td>\n",
       "      <td>2922</td>\n",
       "      <td>2989</td>\n",
       "      <td>6810</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>36472219</td>\n",
       "      <td>36464177</td>\n",
       "      <td>1910</td>\n",
       "      <td>2923</td>\n",
       "      <td>2990</td>\n",
       "      <td>6805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>143</th>\n",
       "      <td>34254915</td>\n",
       "      <td>34247263</td>\n",
       "      <td>1910</td>\n",
       "      <td>2923</td>\n",
       "      <td>2990</td>\n",
       "      <td>6418</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>11942317</td>\n",
       "      <td>11934270</td>\n",
       "      <td>1909</td>\n",
       "      <td>2922</td>\n",
       "      <td>2989</td>\n",
       "      <td>6813</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Start        End  ORF1 Start  ORF1 End  ORF2 Start  ORF2 End\n",
       "132   11706247   11714295        1908      2921        2988      6812\n",
       "133   23237515   23245560        1907      2920        2987      6811\n",
       "134  142476843  142484883        1909      2919        2986      6810\n",
       "135  155523050  155515003        1909      2922        2989      6813\n",
       "136  141428248  141420202        1908      2921        2988      6812\n",
       "137   96064845   96056798        1909      2922        2989      6813\n",
       "138   83066639   83058591        1909      2922        2989      6813\n",
       "139   64020288   64012236        1909      2922        2989      6813\n",
       "140   54125747   54117700        1909      2922        2989      6813\n",
       "141   47790701   47782657        1909      2922        2989      6810\n",
       "142   36472219   36464177        1910      2923        2990      6805\n",
       "143   34254915   34247263        1910      2923        2990      6418\n",
       "144   11942317   11934270        1909      2922        2989      6813"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# include the following headings in the sliced object\n",
    "headings = ['Start', 'End', 'ORF1 Start', 'ORF1 End', 'ORF2 Start', 'ORF2 End']\n",
    "\n",
    "chr_dict = slice_df_by_chrom(l1base_df, headings)\n",
    "\n",
    "ch='X' # ----- Note: this sets 'ch' for the entire noteboook. \n",
    "\n",
    "print ('Active L1s in chromosome {n}: '.format(n=ch))\n",
    "L1chr = chr_dict[ch]\n",
    "L1chr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "177264045 chr2:177263545-177265119 0.01 0.01\n",
      "177265557 chr2:177263545-177265119 0.02 0.03\n",
      "177265176 chr2:177263545-177265119 0.08 0.08\n",
      "177265102 chr2:177263545-177265119 0.02 0.03\n",
      "177264271 chr2:177263545-177265119 0.01 0.02\n",
      "177264953 chr2:177263545-177265119 0.01 0.03\n",
      "177265172 chr2:177263545-177265119 0.06 0.07\n",
      "177230898 chr2:177212457-177213487 0.94 0.95\n",
      "177264039 chr2:177263545-177265119 0.01 0.01\n",
      "177265185 chr2:177263545-177265119 0.03 0.04\n",
      "177265388 chr2:177263545-177265119 0.04 0.05\n",
      "177264609 chr2:177263545-177265119 0.07 0.1\n",
      "177243172 chr2:177263545-177265119 0.92 0.81\n"
     ]
    }
   ],
   "source": [
    "gene_probes = anno.get_probe_refs_by_gene('NFE2L2')\n",
    "for probe in gene_probes:\n",
    "    print (anno.get_probe(probe).hg38_cord, \n",
    "           anno.get_probe(probe).GDC_CGI_Coordinate,\n",
    "           round(anno.get_probe(probe).Beta_Normal, 2),\n",
    "           round(anno.get_probe(probe).Beta_Tumor, 2),\n",
    "          ) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "l1base_df = pd.read_excel('../../data/l1base/l1base_hsfli_8438.xlsx')\n",
    "\n",
    "l1base_list = []\n",
    "for index, row in l1base_df.iterrows():\n",
    "    l1base_list.append([row[0], row[1], row[2], row[5], row[6], row[7], row[8], row[9]])\n",
    "\n",
    "headers = ['Chr','Start','End','Strand','ORF1 Start','ORF1 End','ORF2 Start','ORF2 End']\n",
    "l1base_df = pd.DataFrame(l1base_list, columns=headers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %load '../methylator/annotation/annotate_methyl_450k.py'\n",
    "import os\n",
    "import _pickle as pickle\n",
    "\n",
    "class Probe:\n",
    "    \"\"\"\n",
    "    Holds Illumina 450k probe data with GDC data for a single probe loci.\n",
    "    \"\"\"\n",
    "    def __init__(self):\n",
    "        self.id = None          # REF_ID (Illumina)\n",
    "        self.name = None        # Name (Illumina)\n",
    "        self.chr = None         # Chromosome (Illumina)\n",
    "        self.gene = None        # UCSC_RefGene_Name (Illumina)\n",
    "        self.refseq = None      # UCSC_RefGene_Accession (Illumina)\n",
    "        self.strand = None      # Strand orientation (-/+) (Illumina)\n",
    "        self.seq = None         # Sequence prior to bi-sulfite conv. (Illumina)\n",
    "        self.feature = None             # Defines a UCSC_RefGene_Group feature. (Illumina)\n",
    "        self.tour = None                # Location of CpG relative to CpG island (UCSC_CpG_Island).\n",
    "        self.hg19_cord = None           # Coordinate of probe in hg19. (Ilumina)\n",
    "        self.hg38_cord = None           # Coordinate of probe in hg38. (GDC)\n",
    "        self.GDC_CGI_Coordinate = None  # \n",
    "        self.GDC_gene = None            # Gene name. (GDC)\n",
    "        self.GDC_gene_type = None       # Gene function i.e. protein-coding, etc.\n",
    "        self.Beta_Normal =  None        # Methylation beta-value for normal sample.\n",
    "        self.Beta_Tumor = None         # Methylation beta-value for tumor sample. \n",
    "\n",
    "class Interval:\n",
    "    \"\"\"\n",
    "    Define a genomic interval by chromsome and orientation.\n",
    "    \"\"\"\n",
    "    def __init__(self, chromosome, start, end, strand):\n",
    "        self.chr = chromosome\n",
    "        self.start = start\n",
    "        self.end = end\n",
    "        self.strand = strand\n",
    "\n",
    "class UCSC_RefGene_Group:\n",
    "    \"\"\"\n",
    "    Defines a UCSC_RefGene_Group feature.\n",
    "        • TSS200 = 0–200 bases upstream of TSS.\n",
    "        • TSS1500 = 200–1500 bases upstream of TSS.\n",
    "        • 5'UTR = Between TSS and the ATG start site, in UTR.\n",
    "        • Body = Between ATG and stop codon.\n",
    "        • 3'UTR = Between the stop codon and poly A signal.\n",
    "    \"\"\"\n",
    "    BODY    = \"Body\"\n",
    "    TSS200  = \"TSS200\"\n",
    "    TSS1500 = \"TSS1500\"\n",
    "    UTR5   = \"5'UTR\"\n",
    "    UTR3   = \"3'UTR\"\n",
    "    EXON   = \"Exon\"\n",
    "\n",
    "class UCSC_CpG_Island:\n",
    "    \"\"\"\n",
    "    Defines relation to UCSC CpG island.\n",
    "        • Shore = 0–2 kb from island.\n",
    "        • Shelf = 2–4 kb from island.\n",
    "        • N = upstream (5’) of CpG island.\n",
    "        • S = downstream (3’) of CpG island.\n",
    "    \"\"\"\n",
    "    ISLAND = \"Island\"\n",
    "    NSHORE = \"N_Shore\"\n",
    "    SSHORE = \"S_Shore\"\n",
    "    NSHELF = \"N_Shelf\"\n",
    "    SSHELF = \"S_Shelf\"\n",
    "    \n",
    "class Annotate_Methyl_450k(object):\n",
    "    \"\"\"\n",
    "    Load Illumina probe information along with a single paired GDC Experiment.\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, filepath):\n",
    "        try:\n",
    "            with open(filepath, 'rb') as f:\n",
    "                self.probe = pickle.load(f)\n",
    "        except (FileNotFoundError, IOError):\n",
    "            pass\n",
    "\n",
    "    def get_number(self):\n",
    "        \"\"\"\n",
    "        Return total number of probes.\n",
    "        \"\"\"\n",
    "        number = 0\n",
    "        for probe_id in self.probe.keys():\n",
    "            number += 1\n",
    "\n",
    "        return number\n",
    "\n",
    "    def get_probe(self, ref_id):\n",
    "        \"\"\"\n",
    "        Return probe associated with a ref_id.\n",
    "        \"\"\"\n",
    "        try:\n",
    "            probe = self.probe[ref_id]\n",
    "        except Exception as ex:\n",
    "            probe = None\n",
    "            print(\"WARNING: No probe with ref_id of %s found.\" % ref_id)\n",
    "        return probe\n",
    "    \n",
    "    def get_all_probes(self):\n",
    "        \"\"\"\n",
    "        Return list of all probes.\n",
    "        \"\"\"\n",
    "        probe_list = []\n",
    "        for probe in self.probe.keys():\n",
    "            probe_list.append(self.get_probe(probe))\n",
    "        return probe_list\n",
    "    \n",
    "    def get_probes_by_list(self, list_of_ids):\n",
    "        \"\"\"\n",
    "        Return a list of probes from a list of references.\n",
    "        \"\"\"\n",
    "        out_list = []\n",
    "        for probe_id in list_of_ids:\n",
    "            out_list.append(self.get_probe(probe_id))\n",
    "\n",
    "        return out_list\n",
    "    \n",
    "    def get_probe_refs_by_gene(self, gene_name):\n",
    "        \"\"\"\n",
    "        Get all probes associated with a gene.\n",
    "        \"\"\"\n",
    "        probes = {k: self.probe[k] for k in self.probe \n",
    "                  if gene_name in self.probe[k].gene}\n",
    "        return self.get_keys(probes.keys())\n",
    "\n",
    "    def get_keys(self, dict_keys):\n",
    "        \"\"\"\n",
    "        Get Probe ref_id keys.\n",
    "        \"\"\"\n",
    "        l = []\n",
    "        for i in dict_keys:\n",
    "            l.append(i)\n",
    "        return l\n",
    "    \n",
    "    def get_probe_refs_by_location(self, probe_loc):\n",
    "        \"\"\"\n",
    "        Get all probe references associated with a genomic location.\n",
    "        \"\"\"\n",
    "        probes = {k: self.probe[k] for k in self.probe if probe_loc in self.probe[k].loc}\n",
    "        return self.get_keys(probes.keys())\n",
    "\n",
    "    def get_probes_by_cpg(self, cpg_loc):\n",
    "        \"\"\"\n",
    "        Get a list probes within a cpg location.\n",
    "        \"\"\"\n",
    "        pass\n",
    "\n",
    "    def get_probes_by_chr(self, chr_loc):\n",
    "        \"\"\"\n",
    "        Get a list of probes within a certain genomic region\n",
    "        \"\"\"\n",
    "        pass\n",
    "\n",
    "    def get_probes_by_chr_and_loc(self, chr_loc):\n",
    "        \"\"\"\n",
    "        Get a list of probes within a certain genomic region\n",
    "        \"\"\"\n",
    "        pass\n",
    "\n",
    "    def get_refid_by_chr_and_loc(self, chr_loc):\n",
    "        \"\"\"\n",
    "        Get a list of probe reference *keys* within a genomic region\n",
    "        \"\"\"\n",
    "        pass\n",
    "\n",
    "    def get_hg18_coord(self, probe):\n",
    "        \"\"\"\n",
    "        Get genomic coordinate of a single probe.\n",
    "        \"\"\"\n",
    "        return probe.hg18_cord\n",
    "\n",
    "    def get_hg38_coord(self, probe):\n",
    "        \"\"\"\n",
    "        Get genomic coordinate of a single probe.\n",
    "        \"\"\"\n",
    "        return probe.hg38_cord\n",
    "    \n",
    "    def get_sorted_probes_by_refid(self):\n",
    "        \"\"\"\n",
    "        Sort probes according to probe id.\n",
    "        \"\"\"\n",
    "        sorted_keys = sorted(list(self.probe.keys()))\n",
    "        return sorted_keys\n",
    "    \n",
    "    def get_sorted_probes_by_chr(self):\n",
    "        \"\"\"\n",
    "        Sort probes according to probe id.\n",
    "        \"\"\"\n",
    "        return sorted(self.get_all_probes(), key=lambda x: x.chr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This is not the last line------------------------------------------\n",
    "\n",
    "l = ['cg00000029','cg00000165','cg00000289','cg00000363','cg00000658'] #list of probe refs\n",
    "\n",
    "print( 'Number of probes reported by \\'get_number\\': {a}'.format(\n",
    "    a=anno.get_number()) )\n",
    "print( 'Probe object returned from \\'get_probe\\': {a}'.format(\n",
    "    a=anno.get_probe('cg00000029')))\n",
    "print( 'Length of \\'get_all_probes\\' call: {a}'.format(\n",
    "    a=len(anno.get_all_probes())) )\n",
    "print( 'Number of probes returned by \\'get_probes_by_list\\': {a}'.format(\n",
    "    a=len(anno.get_probes_by_list(l))) )\n",
    "print( 'Get probes associated with a gene (Illumina): {a}'.format(\n",
    "    a=anno.get_probe_refs_by_gene('TP53')) ) \n",
    "print( 'CGI coordinate (GDC): {a}'.format(\n",
    "    a=anno.probe.get('cg00000029').GDC_CGI_Coordinate) )\n",
    "print( 'Gene type (GDC): {a}'.format(\n",
    "    a=anno.probe.get('cg00000029').GDC_gene_type) )\n",
    "print( 'Hg19 coordinate (GDC): {a}'.format(\n",
    "    a=anno.probe.get('cg00000029').hg19_cord) )\n",
    "print( 'Hg38 coordinate (GDC): {a}'.format(\n",
    "    a=anno.probe.get('cg00000029').hg38_cord) )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
